<!DOCTYPE html>
<html xmlns="http://www.w3.org/1999/xhtml" lang="en" xml:lang="en"><head>

<meta charset="utf-8">
<meta name="generator" content="quarto-0.9.446">

<meta name="viewport" content="width=device-width, initial-scale=1.0, user-scalable=yes">

<meta name="author" content="Riya Saju">
<meta name="dcterms.date" content="2022-05-31">

<title>Pasilla_Bulk_RNA</title>
<style>
code{white-space: pre-wrap;}
span.smallcaps{font-variant: small-caps;}
span.underline{text-decoration: underline;}
div.column{display: inline-block; vertical-align: top; width: 50%;}
div.hanging-indent{margin-left: 1.5em; text-indent: -1.5em;}
ul.task-list{list-style: none;}
pre > code.sourceCode { white-space: pre; position: relative; }
pre > code.sourceCode > span { display: inline-block; line-height: 1.25; }
pre > code.sourceCode > span:empty { height: 1.2em; }
.sourceCode { overflow: visible; }
code.sourceCode > span { color: inherit; text-decoration: inherit; }
div.sourceCode { margin: 1em 0; }
pre.sourceCode { margin: 0; }
@media screen {
div.sourceCode { overflow: auto; }
}
@media print {
pre > code.sourceCode { white-space: pre-wrap; }
pre > code.sourceCode > span { text-indent: -5em; padding-left: 5em; }
}
pre.numberSource code
  { counter-reset: source-line 0; }
pre.numberSource code > span
  { position: relative; left: -4em; counter-increment: source-line; }
pre.numberSource code > span > a:first-child::before
  { content: counter(source-line);
    position: relative; left: -1em; text-align: right; vertical-align: baseline;
    border: none; display: inline-block;
    -webkit-touch-callout: none; -webkit-user-select: none;
    -khtml-user-select: none; -moz-user-select: none;
    -ms-user-select: none; user-select: none;
    padding: 0 4px; width: 4em;
    color: #aaaaaa;
  }
pre.numberSource { margin-left: 3em; border-left: 1px solid #aaaaaa;  padding-left: 4px; }
div.sourceCode
  {   }
@media screen {
pre > code.sourceCode > span > a:first-child::before { text-decoration: underline; }
}
code span.al { color: #ff0000; font-weight: bold; } /* Alert */
code span.an { color: #60a0b0; font-weight: bold; font-style: italic; } /* Annotation */
code span.at { color: #7d9029; } /* Attribute */
code span.bn { color: #40a070; } /* BaseN */
code span.bu { } /* BuiltIn */
code span.cf { color: #007020; font-weight: bold; } /* ControlFlow */
code span.ch { color: #4070a0; } /* Char */
code span.cn { color: #880000; } /* Constant */
code span.co { color: #60a0b0; font-style: italic; } /* Comment */
code span.cv { color: #60a0b0; font-weight: bold; font-style: italic; } /* CommentVar */
code span.do { color: #ba2121; font-style: italic; } /* Documentation */
code span.dt { color: #902000; } /* DataType */
code span.dv { color: #40a070; } /* DecVal */
code span.er { color: #ff0000; font-weight: bold; } /* Error */
code span.ex { } /* Extension */
code span.fl { color: #40a070; } /* Float */
code span.fu { color: #06287e; } /* Function */
code span.im { } /* Import */
code span.in { color: #60a0b0; font-weight: bold; font-style: italic; } /* Information */
code span.kw { color: #007020; font-weight: bold; } /* Keyword */
code span.op { color: #666666; } /* Operator */
code span.ot { color: #007020; } /* Other */
code span.pp { color: #bc7a00; } /* Preprocessor */
code span.sc { color: #4070a0; } /* SpecialChar */
code span.ss { color: #bb6688; } /* SpecialString */
code span.st { color: #4070a0; } /* String */
code span.va { color: #19177c; } /* Variable */
code span.vs { color: #4070a0; } /* VerbatimString */
code span.wa { color: #60a0b0; font-weight: bold; font-style: italic; } /* Warning */
</style>

<script src="rna_seq_files/libs/clipboard/clipboard.min.js"></script>
<script src="rna_seq_files/libs/quarto-html/quarto.js"></script>
<script src="rna_seq_files/libs/quarto-html/popper.min.js"></script>
<script src="rna_seq_files/libs/quarto-html/tippy.umd.min.js"></script>
<script src="rna_seq_files/libs/quarto-html/anchor.min.js"></script>
<link href="rna_seq_files/libs/quarto-html/tippy.css" rel="stylesheet">
<link id="quarto-text-highlighting-styles" href="rna_seq_files/libs/quarto-html/quarto-syntax-highlighting.css" rel="stylesheet">
<script src="rna_seq_files/libs/bootstrap/bootstrap.min.js"></script>
<link href="rna_seq_files/libs/bootstrap/bootstrap-icons.css" rel="stylesheet">
<link href="rna_seq_files/libs/bootstrap/bootstrap.min.css" rel="stylesheet">

<script src="rna_seq_files\libs/htmlwidgets-1.5.4/htmlwidgets.js"></script>
<link href="rna_seq_files\libs/datatables-css-0.0.0/datatables-crosstalk.css" rel="stylesheet">
<script src="rna_seq_files\libs/datatables-binding-0.23/datatables.js"></script>
<script src="rna_seq_files\libs/jquery-3.6.0/jquery-3.6.0.min.js"></script>
<link href="rna_seq_files\libs/dt-core-1.11.3/css/jquery.dataTables.min.css" rel="stylesheet">
<link href="rna_seq_files\libs/dt-core-1.11.3/css/jquery.dataTables.extra.css" rel="stylesheet">
<script src="rna_seq_files\libs/dt-core-1.11.3/js/jquery.dataTables.min.js"></script>
<link href="rna_seq_files\libs/crosstalk-1.2.0/css/crosstalk.min.css" rel="stylesheet">
<script src="rna_seq_files\libs/crosstalk-1.2.0/js/crosstalk.min.js"></script>


</head>

<body>

<div id="quarto-content" class="page-columns page-rows-contents page-layout-article toc-left">
<div id="quarto-sidebar-toc-left" class="sidebar toc-left">
  <nav id="TOC" role="doc-toc">
    <h2 id="toc-title">Contents</h2>
   
  <ul>
  <li><a href="#installation-and-loading-of-required-packages" id="toc-installation-and-loading-of-required-packages" class="nav-link active" data-scroll-target="#installation-and-loading-of-required-packages"> <span class="header-section-number">1</span> Installation and loading of required packages</a></li>
  <li><a href="#getting-data-ready-for-the-analysis" id="toc-getting-data-ready-for-the-analysis" class="nav-link" data-scroll-target="#getting-data-ready-for-the-analysis"> <span class="header-section-number">2</span> Getting data ready for the analysis</a></li>
  <li><a href="#starting-from-count-matrix" id="toc-starting-from-count-matrix" class="nav-link" data-scroll-target="#starting-from-count-matrix"> <span class="header-section-number">3</span> Starting from Count matrix</a></li>
  <li><a href="#exploratory-analysis-and-visualization" id="toc-exploratory-analysis-and-visualization" class="nav-link" data-scroll-target="#exploratory-analysis-and-visualization"> <span class="header-section-number">4</span> Exploratory Analysis and Visualization</a>
  <ul class="collapse">
  <li><a href="#pca-plot" id="toc-pca-plot" class="nav-link" data-scroll-target="#pca-plot"> <span class="header-section-number">4.1</span> PCA Plot</a></li>
  <li><a href="#heatmap" id="toc-heatmap" class="nav-link" data-scroll-target="#heatmap"> <span class="header-section-number">4.2</span> 2.Heatmap</a></li>
  </ul></li>
  <li><a href="#differential-expression-analysis" id="toc-differential-expression-analysis" class="nav-link" data-scroll-target="#differential-expression-analysis"> <span class="header-section-number">5</span> Differential Expression Analysis</a>
  <ul class="collapse">
  <li><a href="#constructing-the-deseqdataset-object-and-applying-deseq-function-on-it" id="toc-constructing-the-deseqdataset-object-and-applying-deseq-function-on-it" class="nav-link" data-scroll-target="#constructing-the-deseqdataset-object-and-applying-deseq-function-on-it"> <span class="header-section-number">5.1</span> Constructing the DESeqDataSet object and applying DESeq function on it</a></li>
  <li><a href="#building-the-results-table" id="toc-building-the-results-table" class="nav-link" data-scroll-target="#building-the-results-table"> <span class="header-section-number">5.2</span> Building the results table</a></li>
  <li><a href="#summarizing-the-results" id="toc-summarizing-the-results" class="nav-link" data-scroll-target="#summarizing-the-results"> <span class="header-section-number">5.3</span> Summarizing the results</a></li>
  <li><a href="#annotating-to-other-names-entrezgene-symbol" id="toc-annotating-to-other-names-entrezgene-symbol" class="nav-link" data-scroll-target="#annotating-to-other-names-entrezgene-symbol"> <span class="header-section-number">5.4</span> Annotating to other names-entrez/gene symbol</a></li>
  <li><a href="#plotting-of-deseq2-results" id="toc-plotting-of-deseq2-results" class="nav-link" data-scroll-target="#plotting-of-deseq2-results"> <span class="header-section-number">5.5</span> Plotting of DESEQ2 results</a></li>
  </ul></li>
  </ul>
</nav>
</div>
<div id="quarto-margin-sidebar" class="sidebar margin-sidebar">
</div>
<main class="content" id="quarto-document-content">

<header id="title-block-header" class="quarto-title-block default">
<div class="quarto-title">
<h1 class="title">Pasilla_Bulk_RNA</h1>
</div>



<div class="quarto-title-meta">

    <div>
    <div class="quarto-title-meta-heading">Author</div>
    <div class="quarto-title-meta-contents">
             <p>Riya Saju </p>
          </div>
  </div>
    
    <div>
    <div class="quarto-title-meta-heading">Published</div>
    <div class="quarto-title-meta-contents">
      <p class="date">May 31, 2022</p>
    </div>
  </div>
    
  </div>
  

</header>

<section id="installation-and-loading-of-required-packages" class="level2" data-number="1">
<h2 data-number="1" class="anchored" data-anchor-id="installation-and-loading-of-required-packages"><span class="header-section-number">1</span> Installation and loading of required packages</h2>
<p>Installing the required packages and loading them by using library() in R for bulk RNA Seq Analysis.</p>
<div class="cell">
<details>
<summary>Code</summary>
<div class="sourceCode cell-code" id="cb1"><pre class="sourceCode numberSource r number-lines code-with-copy"><code class="sourceCode r"><span id="cb1-1"><a href="#cb1-1"></a><span class="fu">library</span>(dplyr)</span>
<span id="cb1-2"><a href="#cb1-2"></a><span class="fu">library</span>(textshape)</span>
<span id="cb1-3"><a href="#cb1-3"></a><span class="fu">library</span>(ggplot2)</span>
<span id="cb1-4"><a href="#cb1-4"></a><span class="fu">library</span>(viridis)</span>
<span id="cb1-5"><a href="#cb1-5"></a><span class="fu">library</span>(Rsubread)</span>
<span id="cb1-6"><a href="#cb1-6"></a><span class="fu">library</span>(DESeq2)</span>
<span id="cb1-7"><a href="#cb1-7"></a><span class="fu">library</span>(pheatmap)</span>
<span id="cb1-8"><a href="#cb1-8"></a><span class="fu">library</span>(AnnotationDbi)</span>
<span id="cb1-9"><a href="#cb1-9"></a><span class="fu">library</span>(org.Dm.eg.db)</span>
<span id="cb1-10"><a href="#cb1-10"></a><span class="fu">library</span>(EnhancedVolcano)</span>
<span id="cb1-11"><a href="#cb1-11"></a><span class="fu">library</span>(clusterProfiler)</span>
<span id="cb1-12"><a href="#cb1-12"></a><span class="fu">library</span>(enrichplot)</span>
<span id="cb1-13"><a href="#cb1-13"></a><span class="fu">library</span>(ggplot2)</span>
<span id="cb1-14"><a href="#cb1-14"></a><span class="fu">library</span>(ggridges)</span>
<span id="cb1-15"><a href="#cb1-15"></a><span class="fu">library</span>(europepmc)</span>
<span id="cb1-16"><a href="#cb1-16"></a><span class="fu">library</span>(pathview)</span>
<span id="cb1-17"><a href="#cb1-17"></a><span class="fu">library</span>(biomaRt)</span>
<span id="cb1-18"><a href="#cb1-18"></a><span class="fu">library</span>(dplyr)</span>
<span id="cb1-19"><a href="#cb1-19"></a><span class="fu">library</span>(tidyverse)</span>
<span id="cb1-20"><a href="#cb1-20"></a><span class="fu">library</span>(fgsea)</span>
<span id="cb1-21"><a href="#cb1-21"></a><span class="fu">library</span>(ggnewscale)</span>
<span id="cb1-22"><a href="#cb1-22"></a><span class="fu">library</span>(htmlwidgets)</span>
<span id="cb1-23"><a href="#cb1-23"></a><span class="fu">library</span>(DT)</span>
<span id="cb1-24"><a href="#cb1-24"></a><span class="fu">library</span>(SummarizedExperiment)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</details>
</div>
</section>
<section id="getting-data-ready-for-the-analysis" class="level2" data-number="2">
<h2 data-number="2" class="anchored" data-anchor-id="getting-data-ready-for-the-analysis"><span class="header-section-number">2</span> Getting data ready for the analysis</h2>
<p>Reading sample metadata provided by the client.</p>
<div class="cell">
<details>
<summary>Code</summary>
<div class="sourceCode cell-code" id="cb2"><pre class="sourceCode numberSource r number-lines code-with-copy"><code class="sourceCode r"><span id="cb2-1"><a href="#cb2-1"></a>samples <span class="ot">&lt;-</span> <span class="fu">read.delim</span>(<span class="st">"./data/samples.tsv"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</details>
</div>
<p>Now let’s have a look at the metadata:</p>
<div class="cell">
<details>
<summary>Code</summary>
<div class="sourceCode cell-code" id="cb3"><pre class="sourceCode numberSource r number-lines code-with-copy"><code class="sourceCode r"><span id="cb3-1"><a href="#cb3-1"></a>samples</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</details>
<div class="cell-output cell-output-stdout">
<pre><code>      sample condition        type
1 untreated1 untreated single-read
2 untreated2 untreated single-read
3 untreated3 untreated  paired-end
4 untreated4 untreated  paired-end
5   treated1   treated single-read
6   treated2   treated  paired-end
7   treated3   treated  paired-end</code></pre>
</div>
</div>
<div class="callout-note callout callout-style-default callout-captioned">
<div class="callout-header d-flex align-content-center">
<div class="callout-icon-container">
<i class="callout-icon"></i>
</div>
<div class="callout-caption-container flex-fill">
Making sure the required fields are present in the metadata in the format required for the analysis. Making required changes if needed to make it in the format required.
</div>
</div>
<div class="callout-body-container callout-body">

</div>
</div>
</section>
<section id="starting-from-count-matrix" class="level2" data-number="3">
<h2 data-number="3" class="anchored" data-anchor-id="starting-from-count-matrix"><span class="header-section-number">3</span> Starting from Count matrix</h2>
<ul>
<li><p>Reading counts matrix of the client</p>
<p>We now have all the ingredients to prepare our data object in a form that is suitable for analysis, namely:</p>
<ul>
<li><p>countdata: a table with the fragment counts</p></li>
<li><p>coldata: a table with information about the samples</p></li>
</ul></li>
<li><p>Construction of <em>DESeqDataSet</em> object from the matrix of counts and the sample information table</p></li>
</ul>
<div class="cell">
<details>
<summary>Code</summary>
<div class="sourceCode cell-code" id="cb5"><pre class="sourceCode numberSource r number-lines code-with-copy"><code class="sourceCode r"><span id="cb5-1"><a href="#cb5-1"></a><span class="co">#sample table edits to get the required format</span></span>
<span id="cb5-2"><a href="#cb5-2"></a></span>
<span id="cb5-3"><a href="#cb5-3"></a>samples<span class="sc">$</span>group <span class="ot">&lt;-</span> <span class="fu">paste</span>(samples<span class="sc">$</span>condition ,<span class="st">"_"</span>, samples<span class="sc">$</span>type)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</details>
</div>
<p>Now lets have a look at the edited samples metadata for exploratory analysis:</p>
<div class="cell">
<details>
<summary>Code</summary>
<div class="sourceCode cell-code" id="cb6"><pre class="sourceCode numberSource r number-lines code-with-copy"><code class="sourceCode r"><span id="cb6-1"><a href="#cb6-1"></a>samples</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</details>
<div class="cell-output cell-output-stdout">
<pre><code>      sample condition        type                   group
1 untreated1 untreated single-read untreated _ single-read
2 untreated2 untreated single-read untreated _ single-read
3 untreated3 untreated  paired-end  untreated _ paired-end
4 untreated4 untreated  paired-end  untreated _ paired-end
5   treated1   treated single-read   treated _ single-read
6   treated2   treated  paired-end    treated _ paired-end
7   treated3   treated  paired-end    treated _ paired-end</code></pre>
</div>
<details>
<summary>Code</summary>
<div class="sourceCode cell-code" id="cb8"><pre class="sourceCode numberSource r number-lines code-with-copy"><code class="sourceCode r"><span id="cb8-1"><a href="#cb8-1"></a><span class="co">#reading the counts matrix</span></span>
<span id="cb8-2"><a href="#cb8-2"></a>Counts<span class="ot">&lt;-</span> <span class="fu">read.delim</span>(<span class="st">"./data/counts.tsv"</span>)</span>
<span id="cb8-3"><a href="#cb8-3"></a>countDataMatrix <span class="ot">&lt;-</span> <span class="fu">as.matrix</span>(Counts[ , <span class="sc">-</span><span class="dv">1</span>])</span>
<span id="cb8-4"><a href="#cb8-4"></a><span class="fu">rownames</span>(countDataMatrix) <span class="ot">&lt;-</span> Counts[ , <span class="dv">1</span>]</span>
<span id="cb8-5"><a href="#cb8-5"></a><span class="co">#making a dds object using counts matrix and samples metadata</span></span>
<span id="cb8-6"><a href="#cb8-6"></a>dds <span class="ot">&lt;-</span> <span class="fu">DESeqDataSetFromMatrix</span>(<span class="at">countData =</span> countDataMatrix, <span class="at">colData =</span> samples, <span class="at">design =</span> <span class="sc">~</span> group)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</details>
</div>
</section>
<section id="exploratory-analysis-and-visualization" class="level2" data-number="4">
<h2 data-number="4" class="anchored" data-anchor-id="exploratory-analysis-and-visualization"><span class="header-section-number">4</span> Exploratory Analysis and Visualization</h2>
<p>DESeq2 offers transformations for count data that stabilize the variance across the mean. It minimizes differences between samples for rows with small counts, and normalizes with respect to library size. One such transformation is the regularized-logarithm transformation or rlog2. Another transformation that similarly improves distance calculation across samples, the variance stabilizing transformation. For data sets with larger sample size, the variance stabilizing transformation will be much faster to compute. The two transformations (rlog and VST) are provided for applications other than differential testing. Raw counts work better for differential expression analysis since deseq2 has inbuild normalization. Using rlog transformation in this case.</p>
<div class="cell">
<details>
<summary>Code</summary>
<div class="sourceCode cell-code" id="cb9"><pre class="sourceCode numberSource r number-lines code-with-copy"><code class="sourceCode r"><span id="cb9-1"><a href="#cb9-1"></a><span class="co">#rlog transformation for data visualization</span></span>
<span id="cb9-2"><a href="#cb9-2"></a>rld <span class="ot">&lt;-</span> <span class="fu">rlog</span>(dds, <span class="at">blind=</span><span class="cn">TRUE</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</details>
</div>
<section id="pca-plot" class="level3" data-number="4.1">
<h3 data-number="4.1" class="anchored" data-anchor-id="pca-plot"><span class="header-section-number">4.1</span> PCA Plot</h3>
<p>Principal Components Analysis (PCA) visualizes sample-to-sample distances . In this ordination method, the data points (here, the samples) are projected onto the 2D plane such that they spread out in the two directions that capture the most of the variance across samples. The x-axis is the direction that separates the data points the most. The coordinates of the samples in this direction are called PC1.</p>
<div class="cell">
<details>
<summary>Code</summary>
<div class="sourceCode cell-code" id="cb10"><pre class="sourceCode numberSource r number-lines code-with-copy"><code class="sourceCode r"><span id="cb10-1"><a href="#cb10-1"></a>pca <span class="ot">&lt;-</span><span class="fu">plotPCA</span>(rld, <span class="at">intgroup=</span><span class="st">"group"</span>)</span>
<span id="cb10-2"><a href="#cb10-2"></a>pca</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</details>
<div class="cell-output-display">
<div id="fig-pcaplot" class="quarto-figure quarto-figure-center anchored">
<figure class="figure">
<p><img src="rna_seq_files/figure-html/fig-pcaplot-1.png" class="img-fluid figure-img" width="672"></p>
<p></p><figcaption aria-hidden="true" class="figure-caption">Figure 1: PCA Plot of Sample Distances</figcaption><p></p>
</figure>
</div>
</div>
</div>
<p>In <a href="#fig-pcaplot">Figure&nbsp;1</a>, We can see the distance between the gene expression of samples, investigate the reasons for clustering of samples and if there are any batch effects and add them in deseq2 while modelling if present. PCA is also useful to remove the outlier samples for the downstream analysis .</p>
</section>
<section id="heatmap" class="level3" data-number="4.2">
<h3 data-number="4.2" class="anchored" data-anchor-id="heatmap"><span class="header-section-number">4.2</span> 2.Heatmap</h3>
<p>Sample - sample distance can be viewed in Heatmap also. A heatmap of this distance matrix gives us an overview over similarities and dissimilarities between samples. We have to provide a hierarchical clustering hc to the heatmap function based on the sample distances, or else the heatmap function would calculate a clustering based on the distances between the rows/columns of the distance matrix.</p>
<div class="cell">
<details>
<summary>Code</summary>
<div class="sourceCode cell-code" id="cb11"><pre class="sourceCode numberSource r number-lines code-with-copy"><code class="sourceCode r"><span id="cb11-1"><a href="#cb11-1"></a>rld_mat <span class="ot">&lt;-</span> <span class="fu">assay</span>(rld) </span>
<span id="cb11-2"><a href="#cb11-2"></a>rld_cor <span class="ot">&lt;-</span> <span class="fu">cor</span>(rld_mat) </span>
<span id="cb11-3"><a href="#cb11-3"></a>samples_initial<span class="ot">&lt;-</span> <span class="fu">read.delim</span>(<span class="st">"./data/samples.tsv"</span>)</span>
<span id="cb11-4"><a href="#cb11-4"></a>annot_col <span class="ot">&lt;-</span> samples_initial<span class="sc">%&gt;%</span></span>
<span id="cb11-5"><a href="#cb11-5"></a>  <span class="fu">column_to_rownames</span>(<span class="st">'sample'</span>) <span class="sc">%&gt;%</span></span>
<span id="cb11-6"><a href="#cb11-6"></a>  dplyr<span class="sc">::</span><span class="fu">select</span>(type) <span class="sc">%&gt;%</span></span>
<span id="cb11-7"><a href="#cb11-7"></a>  <span class="fu">as.data.frame</span>()</span>
<span id="cb11-8"><a href="#cb11-8"></a></span>
<span id="cb11-9"><a href="#cb11-9"></a>heatmap<span class="ot">&lt;-</span><span class="fu">pheatmap</span>(<span class="at">mat=</span>rld_cor,<span class="at">main =</span> <span class="st">"Heatmap - Sample Distances"</span>,<span class="at">color =</span> <span class="fu">viridis</span>(<span class="at">n=</span><span class="dv">256</span>),</span>
<span id="cb11-10"><a href="#cb11-10"></a><span class="at">annotation_col =</span> annot_col)</span>
<span id="cb11-11"><a href="#cb11-11"></a>heatmap</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</details>
<div class="cell-output-display">
<div id="fig-heatmap_samples" class="quarto-figure quarto-figure-center anchored">
<figure class="figure">
<p><img src="rna_seq_files/figure-html/fig-heatmap_samples-1.png" class="img-fluid figure-img" width="672"></p>
<p></p><figcaption aria-hidden="true" class="figure-caption">Figure 2: Heatmap Plot of Sample Distances</figcaption><p></p>
</figure>
</div>
</div>
</div>
<p><a href="#fig-heatmap_samples">Figure&nbsp;2</a> helps to find out the relation between the samples,clustering of samples and reasons behind it. PCA plots can be verified based on the results from the heatmap. Use the annotations on top for better analysis of data.</p>
</section>
</section>
<section id="differential-expression-analysis" class="level2" data-number="5">
<h2 data-number="5" class="anchored" data-anchor-id="differential-expression-analysis"><span class="header-section-number">5</span> Differential Expression Analysis</h2>
<p>We can run the differential expression pipeline on the raw counts with a single call to the function DESeq. This function will print out a message for the various steps it performs. The various steps includes:</p>
<ol type="1">
<li><p>The estimation of size factors, controlling for differences in the counts due varying sequencing depth of the samples.</p></li>
<li><p>The estimation of dispersion values. The dispersion parameter captures how much the counts for the samples will vary around an expected value. Note that the expected value takes into consideration the sequencing depth and differences that can be attributed to variables in the design formula.</p></li>
<li><p>Fitting a final generalized linear model using the size factors and dispersion values estimated above, which gives estimates of the log fold changes.</p></li>
</ol>
<section id="constructing-the-deseqdataset-object-and-applying-deseq-function-on-it" class="level3" data-number="5.1">
<h3 data-number="5.1" class="anchored" data-anchor-id="constructing-the-deseqdataset-object-and-applying-deseq-function-on-it"><span class="header-section-number">5.1</span> Constructing the DESeqDataSet object and applying DESeq function on it</h3>
<div class="cell">
<details>
<summary>Code</summary>
<div class="sourceCode cell-code" id="cb12"><pre class="sourceCode numberSource r number-lines code-with-copy"><code class="sourceCode r"><span id="cb12-1"><a href="#cb12-1"></a><span class="co"># Construct the DESeqDataSet object with condition and sequencing type as factors</span></span>
<span id="cb12-2"><a href="#cb12-2"></a>dds_seq <span class="ot">&lt;-</span> <span class="fu">DESeqDataSetFromMatrix</span>(<span class="at">countData =</span> countDataMatrix, <span class="at">colData =</span> samples, <span class="at">design =</span> <span class="sc">~</span>type <span class="sc">+</span> condition) </span>
<span id="cb12-3"><a href="#cb12-3"></a></span>
<span id="cb12-4"><a href="#cb12-4"></a><span class="co">#Saving the dds object as rds in results directory</span></span>
<span id="cb12-5"><a href="#cb12-5"></a><span class="fu">saveRDS</span>(dds_seq,<span class="st">"./results/deseq2/dds.rds"</span>)</span>
<span id="cb12-6"><a href="#cb12-6"></a></span>
<span id="cb12-7"><a href="#cb12-7"></a><span class="co">#filtering the records by condition</span></span>
<span id="cb12-8"><a href="#cb12-8"></a>keep <span class="ot">&lt;-</span> <span class="fu">rowSums</span>(<span class="fu">counts</span>(dds_seq)) <span class="sc">&gt;=</span> <span class="dv">10</span></span>
<span id="cb12-9"><a href="#cb12-9"></a>dds_seq <span class="ot">&lt;-</span> dds_seq[keep, ]</span>
<span id="cb12-10"><a href="#cb12-10"></a></span>
<span id="cb12-11"><a href="#cb12-11"></a><span class="co">#deseq2 analysis using DESeq function</span></span>
<span id="cb12-12"><a href="#cb12-12"></a>dds_seq <span class="ot">&lt;-</span> <span class="fu">DESeq</span>(dds_seq)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</details>
</div>
</section>
<section id="building-the-results-table" class="level3" data-number="5.2">
<h3 data-number="5.2" class="anchored" data-anchor-id="building-the-results-table"><span class="header-section-number">5.2</span> Building the results table</h3>
<p>Calling results without any arguments will extract the estimated log2 fold changes and p values for the last variable in the design formula.This comparison is printed at the top of the output.</p>
<div class="cell">
<details>
<summary>Code</summary>
<div class="sourceCode cell-code" id="cb13"><pre class="sourceCode numberSource r number-lines code-with-copy"><code class="sourceCode r"><span id="cb13-1"><a href="#cb13-1"></a>res <span class="ot">&lt;-</span> <span class="fu">results</span>(dds_seq, <span class="at">contrast =</span> <span class="fu">c</span>(<span class="st">"condition"</span>,  <span class="st">"treated"</span>, <span class="st">"untreated"</span>))</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</details>
</div>
<p>Now lets have a glance of the results table constructed:</p>
<div class="cell">
<details>
<summary>Code</summary>
<div class="sourceCode cell-code" id="cb14"><pre class="sourceCode numberSource r number-lines code-with-copy"><code class="sourceCode r"><span id="cb14-1"><a href="#cb14-1"></a>res</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</details>
<div class="cell-output cell-output-stdout">
<pre><code>log2 fold change (MLE): condition treated vs untreated 
Wald test p-value: condition treated vs untreated 
DataFrame with 9921 rows and 6 columns
              baseMean log2FoldChange     lfcSE       stat    pvalue      padj
             &lt;numeric&gt;      &lt;numeric&gt; &lt;numeric&gt;  &lt;numeric&gt; &lt;numeric&gt; &lt;numeric&gt;
FBgn0000008   95.14429     -0.0405574  0.220040 -0.1843184  0.853764  0.949442
FBgn0000014    1.05652     -0.0834936  2.075676 -0.0402248  0.967914        NA
FBgn0000017 4352.55357     -0.2560572  0.112230 -2.2815484  0.022516  0.130353
FBgn0000018  418.61048     -0.0646154  0.131349 -0.4919353  0.622765  0.859352
FBgn0000024    6.40620      0.3089547  0.755886  0.4087319  0.682736  0.887744
...                ...            ...       ...        ...       ...       ...
FBgn0261570 3208.38861      0.2589831  0.105127   2.463526 0.0137578 0.0908776
FBgn0261572    6.19719     -0.9754562  0.807343  -1.208229 0.2269590 0.5588345
FBgn0261573 2240.97951     -0.0107132  0.101906  -0.105128 0.9162740 0.9704006
FBgn0261574 4857.68037      0.0719383  0.121127   0.593908 0.5525735 0.8256556
FBgn0261575   10.68252      0.5731014  0.747158   0.767042 0.4430565 0.7519263</code></pre>
</div>
</div>
<p>res carries metadata with information on the meaning of the columns</p>
<div class="cell">
<details>
<summary>Code</summary>
<div class="sourceCode cell-code" id="cb16"><pre class="sourceCode numberSource r number-lines code-with-copy"><code class="sourceCode r"><span id="cb16-1"><a href="#cb16-1"></a><span class="fu">mcols</span>(res, <span class="at">use.names =</span> <span class="cn">TRUE</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</details>
<div class="cell-output cell-output-stdout">
<pre><code>DataFrame with 6 rows and 2 columns
                       type            description
                &lt;character&gt;            &lt;character&gt;
baseMean       intermediate mean of normalized c..
log2FoldChange      results log2 fold change (ML..
lfcSE               results standard error: cond..
stat                results Wald statistic: cond..
pvalue              results Wald test p-value: c..
padj                results   BH adjusted p-values</code></pre>
</div>
</div>
<p>Lets have a look at what each column means :</p>
<ol type="1">
<li><p>Base Mean : average of the normalized count values, dividing by size factors, taken over all samples in the DESeqDataSet</p></li>
<li><p>log2FoldChange : It is the effect size estimate. It tells us how much the gene’s expression seems to have changed due to the factor. This value is reported on a logarithmic scale to base 2.</p></li>
<li><p>lfcSE : the standard error estimate for the log2 fold change estimate.</p></li>
<li><p>stat : For the Wald test, stat is the Wald statistic</p></li>
<li><p>pvalue : result of hypothesis test for each gene is reported as a p value.For example, the set of genes with adjusted p value less than 0.1 should contain no more than 10% false positives.</p></li>
<li><p>adjusted p value : satisfy the property that thresholding at a specific value defines a set of tests (one for each gene) with a bounded false discovery rate (FDR), typically a useful metric for assessing which genes to target for further analysis</p></li>
</ol>
</section>
<section id="summarizing-the-results" class="level3" data-number="5.3">
<h3 data-number="5.3" class="anchored" data-anchor-id="summarizing-the-results"><span class="header-section-number">5.3</span> Summarizing the results</h3>
<p>Summarizing the results which will report some additional information</p>
<div class="cell">
<details>
<summary>Code</summary>
<div class="sourceCode cell-code" id="cb18"><pre class="sourceCode numberSource r number-lines code-with-copy"><code class="sourceCode r"><span id="cb18-1"><a href="#cb18-1"></a><span class="fu">summary</span>(res)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</details>
<div class="cell-output cell-output-stdout">
<pre><code>
out of 9921 with nonzero total read count
adjusted p-value &lt; 0.1
LFC &gt; 0 (up)       : 619, 6.2%
LFC &lt; 0 (down)     : 724, 7.3%
outliers [1]       : 0, 0%
low counts [2]     : 1347, 14%
(mean count &lt; 5)
[1] see 'cooksCutoff' argument of ?results
[2] see 'independentFiltering' argument of ?results</code></pre>
</div>
</div>
</section>
<section id="annotating-to-other-names-entrezgene-symbol" class="level3" data-number="5.4">
<h3 data-number="5.4" class="anchored" data-anchor-id="annotating-to-other-names-entrezgene-symbol"><span class="header-section-number">5.4</span> Annotating to other names-entrez/gene symbol</h3>
<p>Our result table so far only contains information about Ensembl gene IDs, but alternative gene names may be more informative for analysis. Bioconductor’s annotation packages help with mapping various ID schemes to each other.</p>
<p>Using the mapIds function to add individual columns to our results table. Providing the row names of our results table as a key, and specify that keytype=ENSEMBL. The column argument tells the mapIds function which information we want, and the multiVals argument tells the function what to do if there are multiple possible values for a single input value. Here just giving us back the first one that occurs in the database.To add the gene symbol and Entrez ID, we called mapIds twice.</p>
<div class="cell">
<details>
<summary>Code</summary>
<div class="sourceCode cell-code" id="cb20"><pre class="sourceCode numberSource r number-lines code-with-copy"><code class="sourceCode r"><span id="cb20-1"><a href="#cb20-1"></a><span class="co">#To get a list of all available key types</span></span>
<span id="cb20-2"><a href="#cb20-2"></a><span class="co">#columns(org.Dm.eg.db)</span></span>
<span id="cb20-3"><a href="#cb20-3"></a> </span>
<span id="cb20-4"><a href="#cb20-4"></a><span class="co">#mapping ensembl id to enrez id and gene symbol using mapIds</span></span>
<span id="cb20-5"><a href="#cb20-5"></a>ids <span class="ot">&lt;-</span> <span class="fu">strsplit</span>(<span class="fu">rownames</span>(res), <span class="st">"."</span>, <span class="at">fixed =</span> <span class="cn">TRUE</span>)</span>
<span id="cb20-6"><a href="#cb20-6"></a>ids <span class="ot">&lt;-</span> <span class="fu">sapply</span>(ids, head, <span class="at">n =</span> <span class="dv">1</span>)</span>
<span id="cb20-7"><a href="#cb20-7"></a><span class="co">#for symbol</span></span>
<span id="cb20-8"><a href="#cb20-8"></a>res<span class="sc">$</span>symbol <span class="ot">&lt;-</span> <span class="fu">mapIds</span>(</span>
<span id="cb20-9"><a href="#cb20-9"></a>  <span class="at">x =</span> org.Dm.eg.db,</span>
<span id="cb20-10"><a href="#cb20-10"></a>  <span class="at">keys =</span> ids,</span>
<span id="cb20-11"><a href="#cb20-11"></a>  <span class="at">column =</span> <span class="st">"SYMBOL"</span>,</span>
<span id="cb20-12"><a href="#cb20-12"></a>  <span class="at">keytype =</span> <span class="st">"ENSEMBL"</span>,</span>
<span id="cb20-13"><a href="#cb20-13"></a>  <span class="at">multiVals =</span> <span class="st">"first"</span></span>
<span id="cb20-14"><a href="#cb20-14"></a>)</span>
<span id="cb20-15"><a href="#cb20-15"></a><span class="co">#for entrez id</span></span>
<span id="cb20-16"><a href="#cb20-16"></a>res<span class="sc">$</span>entrez <span class="ot">&lt;-</span> <span class="fu">mapIds</span>(</span>
<span id="cb20-17"><a href="#cb20-17"></a>  <span class="at">x =</span> org.Dm.eg.db,</span>
<span id="cb20-18"><a href="#cb20-18"></a>  <span class="at">keys =</span> ids,</span>
<span id="cb20-19"><a href="#cb20-19"></a>  <span class="at">column =</span> <span class="st">"ENTREZID"</span>,</span>
<span id="cb20-20"><a href="#cb20-20"></a>  <span class="at">keytype =</span> <span class="st">"ENSEMBL"</span>,</span>
<span id="cb20-21"><a href="#cb20-21"></a>  <span class="at">multiVals =</span> <span class="st">"first"</span></span>
<span id="cb20-22"><a href="#cb20-22"></a>)</span>
<span id="cb20-23"><a href="#cb20-23"></a><span class="co"># Output the first few lines of the results table</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</details>
</div>
<p>Lets have a look at the new results table with additional columns we would love to see:</p>
<div class="cell">
<details>
<summary>Code</summary>
<div class="sourceCode cell-code" id="cb21"><pre class="sourceCode numberSource r number-lines code-with-copy"><code class="sourceCode r"><span id="cb21-1"><a href="#cb21-1"></a><span class="fu">head</span>(res)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</details>
<div class="cell-output cell-output-stdout">
<pre><code>log2 fold change (MLE): condition treated vs untreated 
Wald test p-value: condition treated vs untreated 
DataFrame with 6 rows and 8 columns
              baseMean log2FoldChange     lfcSE       stat    pvalue      padj
             &lt;numeric&gt;      &lt;numeric&gt; &lt;numeric&gt;  &lt;numeric&gt; &lt;numeric&gt; &lt;numeric&gt;
FBgn0000008   95.14429     -0.0405574  0.220040 -0.1843184  0.853764  0.949442
FBgn0000014    1.05652     -0.0834936  2.075676 -0.0402248  0.967914        NA
FBgn0000017 4352.55357     -0.2560572  0.112230 -2.2815484  0.022516  0.130353
FBgn0000018  418.61048     -0.0646154  0.131349 -0.4919353  0.622765  0.859352
FBgn0000024    6.40620      0.3089547  0.755886  0.4087319  0.682736  0.887744
FBgn0000032  989.72022     -0.0483794  0.120853 -0.4003151  0.688924  0.890198
                 symbol      entrez
            &lt;character&gt; &lt;character&gt;
FBgn0000008           a       43852
FBgn0000014       abd-A       42037
FBgn0000017         Abl       45821
FBgn0000018         abo       44793
FBgn0000024         Ace       41625
FBgn0000032      Acph-1       48445</code></pre>
</div>
</div>
<p>The results table can be further reduced to a much better version by giving strict padj and log2FoldChange thresholds</p>
<div class="cell">
<details>
<summary>Code</summary>
<div class="sourceCode cell-code" id="cb23"><pre class="sourceCode numberSource r number-lines code-with-copy"><code class="sourceCode r"><span id="cb23-1"><a href="#cb23-1"></a><span class="co">#Exporting  results </span></span>
<span id="cb23-2"><a href="#cb23-2"></a><span class="fu">write.csv</span>(<span class="fu">as.data.frame</span>(res), <span class="at">file =</span> <span class="st">"./results/deseq2/Treated-Untreated.results.csv"</span>)</span>
<span id="cb23-3"><a href="#cb23-3"></a></span>
<span id="cb23-4"><a href="#cb23-4"></a><span class="co">#filtering the significant genes with padj &lt; 0.05</span></span>
<span id="cb23-5"><a href="#cb23-5"></a>res_sig <span class="ot">&lt;-</span> <span class="fu">subset</span>(res, padj <span class="sc">&lt;</span> <span class="fl">0.05</span>)</span>
<span id="cb23-6"><a href="#cb23-6"></a></span>
<span id="cb23-7"><a href="#cb23-7"></a><span class="co">#filtering again the significant genes with l2fc&gt;1 from res_sig</span></span>
<span id="cb23-8"><a href="#cb23-8"></a>res_lfc<span class="ot">&lt;-</span> <span class="fu">subset</span>(res_sig, <span class="fu">abs</span>(log2FoldChange) <span class="sc">&gt;</span> <span class="dv">1</span>)</span>
<span id="cb23-9"><a href="#cb23-9"></a>res_lfc_new<span class="ot">&lt;-</span><span class="fu">as.data.frame</span>(res_lfc) </span>
<span id="cb23-10"><a href="#cb23-10"></a></span>
<span id="cb23-11"><a href="#cb23-11"></a><span class="fu">write.csv</span>(<span class="fu">as.data.frame</span>(res_sig), <span class="at">file =</span> <span class="st">"./results/deseq2/Treated-Untreated.signif.csv"</span>)</span>
<span id="cb23-12"><a href="#cb23-12"></a><span class="fu">write.csv</span>(<span class="fu">as.data.frame</span>(res_lfc), <span class="at">file =</span> <span class="st">"./results/deseq2/Treated-Untreated.signif-lfc.csv"</span>)</span>
<span id="cb23-13"><a href="#cb23-13"></a></span>
<span id="cb23-14"><a href="#cb23-14"></a><span class="co">#saving the final results to rds object</span></span>
<span id="cb23-15"><a href="#cb23-15"></a><span class="fu">saveRDS</span>(res_lfc,<span class="st">"./results/deseq2/sign_results.rds"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</details>
</div>
<p>Here is the shortlisted differentially expressed significant genes:</p>
<div class="cell">
<details>
<summary>Code</summary>
<div class="sourceCode cell-code" id="cb24"><pre class="sourceCode numberSource r number-lines code-with-copy"><code class="sourceCode r"><span id="cb24-1"><a href="#cb24-1"></a>DT<span class="sc">::</span><span class="fu">datatable</span>(res_lfc_new)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</details>
<div class="cell-output-display">

<div id="htmlwidget-71fd73d417e35989003c" style="width:100%;height:auto;" class="datatables html-widget"></div>
<script type="application/json" data-for="htmlwidget-71fd73d417e35989003c">{"x":{"filter":"none","vertical":false,"data":[["FBgn0000071","FBgn0000078","FBgn0000079","FBgn0000116","FBgn0000406","FBgn0000567","FBgn0001137","FBgn0001224","FBgn0001225","FBgn0001226","FBgn0001967","FBgn0002609","FBgn0002773","FBgn0002868","FBgn0002945","FBgn0003360","FBgn0003366","FBgn0003447","FBgn0003501","FBgn0003748","FBgn0004396","FBgn0010223","FBgn0011260","FBgn0015568","FBgn0016715","FBgn0020248","FBgn0020376","FBgn0023479","FBgn0023549","FBgn0024288","FBgn0024315","FBgn0024836","FBgn0024977","FBgn0024984","FBgn0025111","FBgn0025836","FBgn0026415","FBgn0026562","FBgn0027279","FBgn0027348","FBgn0028373","FBgn0028433","FBgn0028434","FBgn0028542","FBgn0028939","FBgn0029002","FBgn0029090","FBgn0029167","FBgn0029896","FBgn0030041","FBgn0030051","FBgn0030326","FBgn0030598","FBgn0030634","FBgn0030805","FBgn0030954","FBgn0031150","FBgn0031258","FBgn0031322","FBgn0031327","FBgn0031464","FBgn0031547","FBgn0031548","FBgn0031688","FBgn0031689","FBgn0031703","FBgn0031912","FBgn0032004","FBgn0032405","FBgn0032436","FBgn0032498","FBgn0032598","FBgn0032638","FBgn0032808","FBgn0033065","FBgn0033113","FBgn0033363","FBgn0033367","FBgn0033368","FBgn0033446","FBgn0033458","FBgn0033495","FBgn0033541","FBgn0033635","FBgn0033724","FBgn0033764","FBgn0033775","FBgn0033782","FBgn0033857","FBgn0033913","FBgn0034010","FBgn0034139","FBgn0034160","FBgn0034225","FBgn0034290","FBgn0034389","FBgn0034391","FBgn0034403","FBgn0034434","FBgn0034438","FBgn0034636","FBgn0034718","FBgn0034736","FBgn0034751","FBgn0034804","FBgn0034885","FBgn0034897","FBgn0034898","FBgn0035076","FBgn0035085","FBgn0035147","FBgn0035189","FBgn0035266","FBgn0035344","FBgn0035523","FBgn0035539","FBgn0035710","FBgn0035727","FBgn0035765","FBgn0035789","FBgn0035932","FBgn0035957","FBgn0035968","FBgn0035969","FBgn0036181","FBgn0036287","FBgn0036373","FBgn0036560","FBgn0036576","FBgn0036801","FBgn0036821","FBgn0036862","FBgn0036985","FBgn0037046","FBgn0037153","FBgn0037223","FBgn0037290","FBgn0037635","FBgn0037646","FBgn0037678","FBgn0037754","FBgn0038149","FBgn0038198","FBgn0038237","FBgn0038261","FBgn0038351","FBgn0038381","FBgn0038484","FBgn0038528","FBgn0038545","FBgn0038603","FBgn0038832","FBgn0038865","FBgn0039069","FBgn0039094","FBgn0039109","FBgn0039155","FBgn0039178","FBgn0039257","FBgn0039419","FBgn0039464","FBgn0039479","FBgn0039485","FBgn0039593","FBgn0039640","FBgn0039651","FBgn0039759","FBgn0039788","FBgn0039808","FBgn0039827","FBgn0040091","FBgn0040099","FBgn0040391","FBgn0040827","FBgn0041605","FBgn0043364","FBgn0046258","FBgn0050088","FBgn0050148","FBgn0050359","FBgn0050463","FBgn0051092","FBgn0051125","FBgn0051163","FBgn0051462","FBgn0051477","FBgn0051642","FBgn0052135","FBgn0052185","FBgn0052313","FBgn0052407","FBgn0052625","FBgn0052677","FBgn0052681","FBgn0053307","FBgn0053346","FBgn0053643","FBgn0054057","FBgn0062978","FBgn0063498","FBgn0063649","FBgn0063667","FBgn0085334","FBgn0085359","FBgn0086910","FBgn0087007","FBgn0250829","FBgn0250904","FBgn0259224","FBgn0259236","FBgn0259714","FBgn0259715","FBgn0259740","FBgn0259741","FBgn0259935","FBgn0260011","FBgn0260429","FBgn0261015","FBgn0261284","FBgn0261362","FBgn0261546","FBgn0261552"],[342.246487725455,29.0159476197304,159.517803392293,658.401186017792,265.504534086402,261.174489927451,494.04936828311,598.932007347051,420.302629286048,1323.90943619999,14.3462419633866,43.2551136137613,41.1047663249845,100.129906548607,55.48841088399,4342.8320615483,17.7120740521515,36.3051835360252,277.460809952997,1095.97418687562,2423.4177628592,65.4037505036009,211.129678246791,210.367106107003,150.872486038183,35.6263967825634,40.8130277420041,3091.78196454293,216.537978330365,58.8494875253442,92.0840267409837,16.4566488876465,28.443748917727,120.72437320106,1501.4479295742,93.7244693947194,46.9558100716458,43909.4443835314,2954.90726348927,172.975837264644,42.0609200090626,92.7233225298929,83.6654680040866,1164.01227304999,20.0277226360595,1587.74346320704,43.5366755871157,3706.02400785474,489.877246619059,86.8708892913224,32.90043594538,140.61710360818,37.9502625019135,8.035542593361,340.916155958115,17.0091428165611,795.375914895633,107.568117135677,396.807063925681,756.387956217435,29.1644059228605,265.28710570863,53.6644499065255,30.8038354303719,274.174934431934,90.8193992261229,1318.33225228138,961.575715718715,79.6159906120173,17.2796257212951,57.3565674133406,315.324408370831,131.934790702139,28.344772816723,29.9482901809733,373.890809050359,70.2294434729282,15324.9710040484,63.8160509395739,81.54457073364,41.1104671066625,66.4276945513304,22.7480713547379,2490.12064966662,152.834926132809,65.2059280189589,46.8809721686594,162.226101633926,17.5401051662992,5212.96176538917,547.240149120962,28.9000637902582,42.0373719992831,49.100787811625,13.5368226966942,110.747440038619,65.1274308024795,68.0353080384092,114.623344446462,74.5954977692353,15.1659174890807,496.04391902447,225.870671605661,16640.8459837339,39.021954820101,160.525286659183,892.485155473449,6.68186364603847,30.8126152894526,638.219335651327,19322.7013020077,215.654948171641,383.49566591085,77.3889343639722,47.9047895792265,20.4818430733033,17.3496883088748,65.2390054012443,556.565720032329,13.6150910959808,27.3932665883248,14.2090124459709,295.082687909235,12.9148030214194,89.4526680649813,14.1818722989093,3859.03951894242,18.8949531546659,1775.62783509235,33.9254902851723,416.309296705106,261.055497817731,1932.74323423911,24.8266044951045,103.140156366507,24.9877279331403,67.0998167313376,167.519023619718,39.6175652978946,35.3320883204162,269.747253760138,993.47812985692,62.2923420739071,20.7717228830904,29.1342891179027,97.3782746279397,158.231816828124,27.3441989271922,152.60441974042,81.9815873557935,20.075687668467,301.911362068779,282.793771098766,33.7380206190465,15.627548563531,475.403846877735,730.567672556595,13.9248858729795,521.915451167556,1564.28099570834,820.539132138972,175.681250169556,58.5700497791369,42.4951299978344,72.0228298705173,74.2294603344455,14.2508316554665,17.2209151090385,38.2619635193833,261.911169759448,1106.50895128494,910.08257602036,21.726859188949,215.979868086714,40.2580752965441,1194.93765071583,9.75968862871394,14.4111299865994,57.124222972352,45.7477496998988,42.0945642373631,153.068561614784,27.1196387196084,12.1723509831174,90.4142083486468,43.2700083831999,110.54063539156,240.709567638936,23.6343938829458,12.5303002914293,187.050167913661,155.80098746963,46.1447404602111,160.546278938595,1013.56661385984,12.2703655715405,14.3131705367409,30.5141184791153,17.3918126747688,343.868523446746,646.192411101866,13.0521699727757,43.2621295875823,68.6061082175759,87.0684066671532,133.273160266023,76.2545590006507,30.2591383074778,251.947326325558,6.31393317037846,117.021360483194,804.639164765266,46.5222203495388,33.1623213343193,44.8983152304953,89.2793535054463,68.3370323094981,434.523396271828,51.0138514919771,5796.45165295305,28.1213312285079,5128.6158935707],[2.62144935566491,-1.14501695430644,-1.2674098238261,1.19422563140812,-2.07542043020377,-1.22965264298093,-1.22842363076624,1.46821077964139,1.30749131977333,1.66579284946529,-1.98133960079511,-1.46435789196936,-1.20518989722174,-2.04432432936056,-1.03115306661152,-3.12666665609594,-1.89344876799415,1.95400795141459,2.13237206335522,1.10011258911794,1.1134045237883,-1.40409397886532,2.15957715662833,1.06183448335038,-2.00754815661109,2.08899117295138,-1.48856548087101,-1.53218678343173,-1.61771847045338,-4.62135607415508,-2.29394773738732,1.43279825895802,1.03797838802377,1.438534504972,2.85211110513182,-1.04248164123098,1.1432424021751,-2.47776427863145,-1.18260015023724,-1.22687573041698,1.26355664341467,-1.31542019867536,1.02286670050077,-1.05461015846372,-2.44509594143889,1.42278919730895,-1.52699965896887,-2.1970228571432,-2.44290318953614,-2.83494563290541,1.02677699351213,-1.2601317246253,2.1570500748155,-3.20128460589783,1.00062156811489,1.4528082267175,1.2425282147863,-1.30936705624077,-1.48917123314998,1.19517940603363,-1.08459617432371,-1.35408734961225,-1.89492062971288,1.66162437774143,1.00267651053584,-1.85946688485566,-1.11930208028039,-1.14706202362392,2.65454715371732,2.32844752857803,1.06636450700898,1.05305792199133,-1.38668102055061,-1.17476173318497,2.96957734353858,1.04761808587662,-1.00805297479342,1.37999201896334,1.31468870232389,-1.0931364325351,1.14273902288232,-2.02345379349732,1.13677990453948,1.18569618711508,1.40117632985879,3.41717552365792,1.93850360435856,-1.43067641449812,1.80149756690393,-1.16949417312007,-1.46291456944082,1.21063230569018,-1.01605203884835,-1.31402338316681,-1.83804752125488,-1.91506295082325,1.51983852752286,1.06102765814907,-3.65188054254781,1.91367550484461,-2.36675028396253,1.34895524425252,-3.53495293105837,-1.08431052543867,-1.47570785644026,1.56631246692317,-1.93388649654442,-2.77022667865249,1.3114531164104,-2.58312818614388,-1.17666381496019,2.89094853309672,1.00689615438268,-1.65921659580352,1.60925772366893,1.76266088280413,2.32561805807725,1.56475307975694,1.02855321782975,-2.5932162424813,1.50042062365891,1.74201913540907,1.09844667239749,1.70624074657628,-1.0127749936644,1.66267700823862,-1.06771481651447,2.26569350641701,-1.10564353683322,-1.05752430303921,-1.74903445308743,1.07216434914196,-1.33270490893099,1.61011544621266,1.80265154513008,-2.60995223829902,3.08217867091789,-1.09401355679272,-2.79574395967225,1.60955331054492,-2.25909779712478,1.01382795502769,2.39391537571597,-2.87155915084452,1.40927243862408,-1.63193415265173,-1.13308154409601,-1.3732487751445,1.3833401108224,1.11331172029282,-1.28455133548593,-2.49369632157339,-1.04118591532397,-1.04405104705798,-1.77910826548052,-1.28531956826896,-4.61978906928183,2.17441979602392,-1.0474351231843,-1.16698382338491,1.03894022637995,1.47949406552964,1.34697792402423,1.97477081884836,-1.80456720466175,1.07195993819758,1.9061407746145,1.8950746369695,-1.01502025486328,-4.1848561459937,-1.57547761799752,-1.74768484599517,-1.32858936880275,-1.97460891807196,1.02736340731909,-1.14883377939173,2.0894165184901,-1.51270118714549,-1.08482372232659,1.10563693737662,2.66625636887581,2.31728419897246,1.28229967765935,1.71025104548105,1.20372135080345,1.14047261818986,1.96842467409999,1.39909460978054,1.35098363107967,-1.84911606462849,-2.17063319136058,1.63804427546305,1.20582156843179,-1.00135866164269,-1.40593515585827,1.87948166341236,1.71621909073618,-1.0435952719003,-1.61942990618212,-1.34015709288962,-1.29463974793089,2.73484958449767,-1.29148308021932,-4.89695065377064,1.73554553102418,1.10843330088809,1.20066359582718,1.77832452793955,1.11055960205708,-3.0254320295889,-1.03799634081773,-1.8642734173464,1.39721468559686,1.35148185871162,1.14835195810653,2.29976758117281,-1.36714543814068,-1.08608079777222,2.30722488247242,1.49116696984791,-1.86830269215088,-1.86872895974749],[0.157981395703166,0.376028759374252,0.184194419695086,0.144103573597402,0.222764047437841,0.165886864751544,0.143565951501735,0.134219843524878,0.134581450621955,0.131238077578343,0.563402488052751,0.33941573723242,0.340121108099814,0.24121667462734,0.322384856436554,0.109269080745043,0.514557247837783,0.35579331252669,0.172458074882153,0.101655389913463,0.153260136420748,0.263121490848304,0.206173171252189,0.166443053144263,0.199358455621803,0.351073945547538,0.353747183379701,0.124905051918807,0.190161575389267,0.468074233428786,0.244437032316555,0.507036418059598,0.376517158336688,0.198159314591089,0.104609098358856,0.233171784807184,0.294468890919974,0.0883064271341897,0.0930329784028638,0.201208965730702,0.35179557570127,0.225108399626727,0.237567601135731,0.151726088932886,0.515662707264745,0.327819175497064,0.317240913964745,0.0967652140406412,0.147401040440381,0.32433919187043,0.347360327520132,0.228489171367704,0.349622481528805,0.932795948719653,0.132886632294799,0.478435832939038,0.109291676957692,0.2408441159083,0.153597317909437,0.138657116386215,0.381560863568815,0.206936849845358,0.323182470812219,0.384956265395038,0.147219556829119,0.262096669380211,0.104171225477662,0.133922834340427,0.257362331864775,0.540170418909829,0.291404906216785,0.138446159611535,0.214185101606452,0.405468262616291,0.436961235514999,0.186778555854678,0.255817805651078,0.157001583544813,0.282970800820836,0.245128880525246,0.338924640410725,0.360141102458716,0.409890290541027,0.140433863141938,0.205959770456353,0.312570474376987,0.310717236580502,0.207821936335752,0.5380838519643,0.100327102207594,0.132839626162507,0.369186789749647,0.366948276330037,0.312134266004859,0.60230554348047,0.249905811664595,0.266012608748334,0.274986617319557,0.276660243691908,0.258192402951486,0.63410651470379,0.191422215127145,0.20475119458336,0.103543829818166,0.363727985913044,0.19550624329186,0.140748408160615,0.925307606617926,0.373619422954941,0.129240918382656,0.108985882059896,0.215806012838373,0.15196415692131,0.249102915682389,0.356081289450525,0.500555703830809,0.547701422137639,0.257584095974868,0.122595398878901,0.761580246852268,0.405691540870431,0.571928969109684,0.143244660113858,0.587087997650618,0.274132609667273,0.58968665269815,0.141061237919948,0.504198321974425,0.15162361780694,0.360472391699035,0.472712535215823,0.38690209648345,0.149235737700532,0.425972096465988,0.218418121140833,0.473334939439531,0.321459602300154,0.175573457373854,0.399735425259903,0.349927005635479,0.174274523183455,0.10955974803118,0.30803720662934,0.566834593580085,0.380206532797056,0.251697744224763,0.191156342242472,0.420349146471706,0.188147976265332,0.240689620042402,0.454661223650297,0.174818507832282,0.155222676119166,0.359628757088706,0.533836859251181,0.128436819801663,0.166467360709367,0.59563822142855,0.123604517601947,0.0985621212331464,0.116558470229177,0.171723915106622,0.27576274091169,0.33265214300746,0.270403042898132,0.246145408922327,0.56630411511932,0.501862913264731,0.349441464532836,0.21618142660635,0.117218977132138,0.139465654318034,0.451012031844725,0.175153365998661,0.356110330565071,0.158526682988533,0.660860796436108,0.550819659207491,0.334639684771744,0.31112273482481,0.358085378731037,0.188734099825502,0.382255128429304,0.569504264401759,0.293639117267654,0.318484067906773,0.230286315519007,0.156282313062298,0.414292915382862,0.641644753509053,0.184199089742156,0.213940187336484,0.293562252776066,0.182865181945757,0.152116699424055,0.615684565232891,0.555488280603118,0.376529891657299,0.502683021923958,0.148154099291927,0.123936274572261,0.612739190363545,0.395067664797987,0.478349974322431,0.235624943358,0.219995240406458,0.279976529714102,0.370274876286861,0.185053258933513,1.02899738382945,0.232653030385009,0.306444916786802,0.456181273526282,0.355436056358373,0.303880053105324,0.244372325525986,0.365990931708527,0.156382107742368,0.310706539001826,0.145924732796168,0.417268455517766,0.146616832085609],[16.5934054702897,-3.04502495025078,-6.88082638944308,8.28727283852487,-9.31667589126064,-7.41259800661512,-8.5565109130447,10.9388503300501,9.71524168992746,12.6929080355577,-3.51673917458739,-4.31434883930152,-3.54341400319105,-8.47505394276279,-3.19851582983537,-28.6143768646811,-3.67976308943387,5.49197492650457,12.3645823184119,10.8219799270304,7.26480185774889,-5.33629531490766,10.4745789353298,6.37956624377732,-10.0700426794014,5.95028824965455,-4.20799246130881,-12.2668119495095,-8.50707335139529,-9.87312640625878,-9.38461621648466,2.82582908825615,2.7567890733298,7.25948466233082,27.2644650405819,-4.47087387564081,3.8823877069099,-28.0587082848031,-12.7116230237861,-6.09752018734112,3.59173545857104,-5.84349673693464,4.3055816349148,-6.95075030194847,-4.74165749625087,4.34016465068467,-4.81337555072914,-22.7046762509145,-16.5731746684937,-8.74068168128735,2.95594203529942,-5.51506102929226,6.16965495291751,-3.43192378814668,7.52988882956328,3.03657904925909,11.3689189275346,-5.43657482061678,-9.69529451046802,8.61967591122106,-2.84252468709517,-6.54348102150075,-5.8633150026689,4.316397801803,6.81075620747632,-7.09458418244231,-10.7448297276718,-8.56509667879442,10.3144354283831,4.31057948948277,3.65939105436915,7.60626314912668,-6.47421790848241,-2.89729638912993,6.79597433863593,5.60887774874818,-3.94051138163679,8.78966942756631,4.64602248186127,-4.45943550263355,3.37166109108354,-5.61850280260436,2.77337602468946,8.44309314425597,6.80315542571323,10.9324962009576,6.23880292478182,-6.88414533962744,3.34798667591953,-11.6568120416773,-11.0126369043766,3.27918641539459,-2.76892440812151,-4.20980176250933,-3.05168621001489,-7.66313891648707,5.71340785188395,3.85847016299005,-13.1998746687095,7.41181956931623,-3.73241754986244,7.04701511972645,-17.2646266521253,-10.471995553408,-4.05717435444479,8.01157262576475,-13.7400239321895,-2.99384405665689,3.51013099382834,-19.9869222415749,-10.7964792569512,13.3960518294821,6.62587925193487,-6.6607674633526,4.519354909527,3.52140804572656,4.2461420841314,6.0747270666494,8.38981908975027,-3.40504661616352,3.69842718544165,3.04586623426482,7.6683254476948,2.90627768478361,-3.69447106235792,2.81959410244565,-7.56915813485486,4.49365538850789,-7.29202714474876,-2.9337178862845,-3.69999592308015,2.77115156233801,-8.93019949152724,3.7798613091579,8.25321422835496,-5.51396489215316,9.5880746721014,-6.23108739302892,-6.99398597923738,4.59968303281398,-12.9628689027981,9.25365358397098,7.77151371391494,-5.06595607143164,3.70659711777311,-6.48370591352793,-5.92751216519278,-3.26692414311096,7.35240494360446,4.62550782246735,-2.82529335836642,-14.2644869384528,-6.70769208053491,-2.90313559880433,-3.33268157612065,-10.007407301534,-27.7519211549672,3.65057129948594,-8.47408447122814,-11.8400842918593,8.91346827336695,8.61553887011618,4.88455372749426,5.93644400121623,-6.67362018311891,4.3549865215477,3.36593134982408,3.77608025395145,-2.90469322586042,-19.3580744270598,-13.4404654992129,-12.5312920556755,-2.94579584355781,-11.2735996069128,2.88495816925301,-7.24694264545249,3.16165905098004,-2.74627305300238,-3.24176650795779,3.5537002398722,7.44586773781255,12.2780366723075,3.35456500722005,3.00305221994712,4.0993221952314,3.58094087935257,8.54772750896493,8.9523541235458,3.26093829007715,-2.88183773733981,-11.7841689359,7.65655249654772,4.10754978553597,-5.47593943793916,-9.24247739519345,3.05266977531167,3.08956849435744,-2.771613343384,-3.22157271193275,-9.04569700936143,-10.4460114877509,4.46331755420287,-3.26901742484974,-10.2371713528511,7.36571224714208,5.03844219011365,4.28844373867067,4.80271452865693,6.00129718577985,-2.94017465654738,-4.46156381070982,-6.08355144831266,3.06284971935912,3.80232065524879,3.778964582807,9.41091662577911,-3.73546260219766,-6.94504514264178,7.42573648396521,10.2187404511531,-4.47745969637846,-12.7456645540967],[7.77821632629075e-62,0.00232661007847206,5.95063372725946e-12,1.15874622582648e-16,1.20042235941152e-20,1.23848865691312e-13,1.16334829842535e-17,7.51467640015488e-28,2.5963233723551e-22,6.47366896275328e-37,0.000436882855898548,1.6007401635977e-05,0.000394982106721358,2.34970634478606e-17,0.00138136949006261,4.4509466840057e-180,0.000233450733535679,3.97463831287198e-08,4.06317843781705e-35,2.70861808232485e-27,3.73585475736108e-13,9.48648500572699e-08,1.13039519052432e-25,1.7759025542969e-10,7.49455557856316e-24,2.67670663457948e-09,2.57649460115358e-05,1.36523697797351e-34,1.78378690204666e-17,5.44407971394139e-23,6.31459848194367e-21,0.00471584187430859,0.00583719916458102,3.88567848103201e-13,1.11977397682652e-163,7.7900631211995e-06,0.000103435786044083,3.12773537416913e-173,5.09655404641913e-37,1.07726518788685e-09,0.000328483203903217,5.1116300167728e-09,1.66547556497394e-05,3.63348498536765e-12,2.1197677231767e-06,1.423759950343e-05,1.48402030326362e-06,4.02790333808468e-114,1.0892032281047e-61,2.31708028154681e-18,0.00311715667275661,3.48658948268913e-08,6.84391774571809e-10,0.00059931595115823,5.07835411090019e-14,0.00239279358392354,5.97226855991833e-30,5.4314537371314e-08,3.15725470720801e-22,6.71438773298967e-18,0.00447577646409257,6.01031152207013e-11,4.5371630228751e-09,1.58596057152335e-05,9.70870389901995e-12,1.29740981778239e-12,6.26774654414667e-27,1.07985201534885e-17,6.06366338564879e-25,1.62827283829197e-05,0.000252815294943413,2.82134548627108e-14,9.53042798607209e-11,0.00376393973375644,1.07582928098362e-11,2.0364280401727e-08,8.13080926016836e-05,1.5000067590117e-18,3.3839614118929e-06,8.21757967004785e-06,0.000747163289718758,1.92619271590296e-08,0.00554779746707025,3.09048641902015e-17,1.0235211753338e-11,8.0600225395759e-28,4.40932075640338e-10,5.81354974723493e-12,0.000814009248366134,2.1182571894377e-31,3.32137868702031e-28,0.00104106840966066,0.00562416823348448,2.5559485320827e-05,0.00227559853208954,1.81442902917746e-14,1.10735848659262e-08,0.000114098993842258,8.78811453081201e-40,1.24578176517214e-13,0.000189650783676586,1.82796463722906e-12,8.68637921111584e-67,1.1616787501396e-25,4.96700020570249e-05,1.13250824335265e-15,5.84564667761511e-43,0.00275486708846446,0.000447885987375418,7.15767787972849e-89,3.57656045463577e-27,6.37662048298534e-41,3.45187260001293e-11,2.72401346781592e-11,6.2028338801112e-06,0.000429261517132095,2.17482745996874e-05,1.24198779362951e-09,4.86890828592406e-17,0.000661527602198423,0.00021693956309542,0.00232011029557068,1.74256291511725e-14,0.00365756702833668,0.000220344984349504,0.00480844298541867,3.75650650370447e-14,7.0010892165485e-06,3.05325623479323e-13,0.00334928481287289,0.000215602930571133,0.0055858419957449,4.25241267325797e-19,0.000156915739373171,1.54190295905934e-16,3.50838613892351e-08,8.97439918046655e-22,4.63208596926672e-10,2.67184271925918e-12,4.23134272849491e-06,1.98675566884883e-38,2.16947321165937e-20,7.75536806244907e-15,4.06355625085553e-07,0.000210062705653346,8.94965902622293e-11,3.0755857172295e-09,0.00108722808559906,1.94671972318079e-13,3.73681701929427e-06,0.00472373459351045,3.64261005492783e-46,1.97726603752063e-11,0.00369446636418869,0.000860133234555633,1.41409870754578e-23,1.65190004998792e-169,0.00026165765981138,2.36935467216309e-17,2.42209411890146e-32,4.94605636629722e-19,6.96133993907348e-18,1.03663373980116e-06,2.91270337530621e-09,2.49569020760753e-11,1.33075106024425e-05,0.000762856912998814,0.000159315678433566,0.00367613215322904,1.74279127577011e-83,3.5022652856461e-41,5.03358814438043e-36,0.00322124935474832,1.77175310509836e-29,0.00391465672990697,4.26284368319025e-13,0.00156873122006879,0.00602765730734753,0.00118791305328312,0.000379851903399757,9.63092581012994e-14,1.18847455006095e-34,0.00079489873537259,0.00267286563745671,4.14361860604087e-05,0.000342359122585182,1.25535667737506e-17,3.47980566617083e-19,0.00111044204281893,0.00395363274429062,4.71043357563534e-32,1.90990704441001e-14,3.99878601971404e-05,4.35196523480553e-08,2.40854843527561e-20,0.00226815386640092,0.00200447481609014,0.00557792494141169,0.00127489108462163,1.48713350514146e-19,1.5281754239964e-25,8.07003151752144e-06,0.00107921655190184,1.35129372448119e-24,1.76203533654349e-13,4.69336130725508e-07,1.79929360619899e-05,1.56528971329907e-06,1.95747338096583e-09,0.00328027286097519,8.13637104090061e-06,1.17549173875765e-09,0.00219240131247302,0.000143347029802319,0.000157481818578604,4.91828821101789e-21,0.000187370424944673,3.78339322958357e-12,1.12154089519141e-13,1.63450825651444e-24,7.55364821815337e-06,3.29570384289157e-37],[7.41004742017966e-59,0.0226044025628499,3.67055637248364e-10,1.24188626752953e-14,1.94196628482912e-18,9.8901230135055e-12,1.42493547295699e-15,2.08685825262443e-25,4.83932099882014e-20,2.92132829929719e-34,0.00555761662681625,0.000323696843459592,0.00513896294845057,2.70864626121685e-15,0.0146429002692328,3.81624168686649e-176,0.00331391819426309,1.29085412479411e-06,1.65893771075445e-32,7.03748225389493e-25,2.78532336431425e-11,2.97938177432612e-06,2.69195502802621e-23,8.36625741788e-09,1.49437952396745e-21,1.04674174582139e-07,0.000483164331465915,5.08936602136734e-32,2.09509436960933e-15,1.03727865482963e-20,1.04118014200356e-18,0.0399147366538222,0.0471263141592445,2.87205235313521e-11,2.40023551932764e-160,0.000173937503128033,0.00168604264171477,1.34086015490631e-169,2.42765857744431e-34,4.48372413637952e-08,0.0044309480998774,1.90552677233956e-07,0.000333639894721648,2.37812979118643e-10,5.40919299360626e-05,0.000291344100578541,3.87926526834826e-05,6.90704864414761e-111,9.33882847776968e-59,3.15343592602894e-16,0.0288311772515806,1.15868287692157e-06,2.93398753758935e-08,0.00730943807287435,4.31107011355032e-12,0.0230514743691691,1.89652706047184e-27,1.71842377646364e-06,5.75963869353222e-20,8.59241200338111e-16,0.0381843854757509,3.0313182935429e-09,1.70621209465487e-07,0.000321466334284662,5.78072411320813e-10,9.19338163443486e-11,1.53541882484324e-24,1.34183350429001e-15,1.33307307355263e-22,0.000327718575481581,0.00354189271053075,2.44345618174629e-12,4.61660392952441e-09,0.0334425070230339,6.31791798298187e-10,7.02523844545244e-07,0.0013669325215036,2.07436418576875e-16,8.33738078895681e-05,0.00018206079610075,0.00884831221829921,6.68630621301698e-07,0.045431533412283,3.48655665219457e-15,6.05218659124967e-10,2.15958228919762e-25,1.92885286558176e-08,3.61198373426031e-10,0.00952157611936048,6.98536043932262e-29,9.81982788362487e-26,0.0116834038539666,0.0457945094338993,0.000481641817891805,0.0221967938727369,1.62051192668412e-12,4.02308968815471e-07,0.00181499957922732,5.02328626581215e-37,9.8901230135055e-12,0.0028035617573156,1.25383750396816e-10,9.3096269195134e-64,2.69195502802621e-23,0.000876276949870229,1.15596734267924e-13,4.176714551156e-40,0.0258427028626852,0.00568073144342726,1.0228321690132e-85,9.01924392295502e-25,3.90522457293688e-38,1.81572734187183e-09,1.4689114133996e-09,0.000142965316365789,0.00549326604162773,0.000417158179905414,5.11961699162471e-08,5.42156099266401e-15,0.00797740880625778,0.00313138015821571,0.0225795978141011,1.57270888781214e-12,0.0327690487993299,0.00316986224129639,0.0404986150854417,3.22082867627621e-12,0.000159369333705846,2.33737669259975e-11,0.030484891704429,0.00311733478367099,0.0455789399002893,6.17969258652777e-17,0.00237702393884376,1.63213283592281e-14,1.16142481680039e-06,1.60305205361084e-19,2.01601548733466e-08,1.78971714647876e-10,0.000102195866349621,1.06465269404437e-35,3.44464135495693e-18,7.30709074367454e-13,1.16136437649451e-05,0.00305785677125941,4.38482151376202e-09,1.17200319731226e-07,0.0120437901885353,1.50370945104073e-11,9.10212191006508e-05,0.0399421108528192,2.83924896463193e-43,1.10837863429902e-09,0.0328764608861321,0.0100200847188587,2.75556416329489e-21,4.72113034286547e-166,0.00364196879094606,2.70864626121685e-15,8.65293123977547e-30,7.06791454743873e-17,8.77743068200236e-16,2.82161831271592e-05,1.11988873272984e-07,1.36293298344121e-09,0.000276267786695743,0.00900927709649013,0.00240065488029771,0.0328168715143259,2.13467034263613e-80,2.30987865839459e-38,2.15789923749589e-33,0.0294759786207173,5.42536111539762e-27,0.0344364502946089,3.12389929398916e-11,0.0162443254599877,0.0483396691917597,0.0129253382218902,0.00497989330236929,7.9399574900052e-12,4.63180945101027e-32,0.00932347709587495,0.0253228176525457,0.000740153873504051,0.00456455883244397,1.51597579603011e-15,5.14411272099115e-17,0.0122534492601409,0.0345902521934161,1.6154902990999e-29,1.6882003091517e-12,0.000720285532206474,1.40806603483859e-06,3.75470805164602e-18,0.0221519614736962,0.0200307308544952,0.0455789399002893,0.0137447700797784,2.23696187247067e-17,3.44804633824872e-23,0.000179720649951244,0.0119860138808372,2.89649809842543e-22,1.37342645232035e-11,1.32808184318169e-05,0.00035710980045255,4.07926869356419e-05,7.84714263965363e-08,0.0299840719722829,0.000180728614778968,4.8689208541585e-08,0.0215076073834596,0.00219083321483972,0.00238047334788062,8.26851041593478e-19,0.00277463561912889,2.45748587503405e-10,9.0717845616709e-12,3.41811555886704e-22,0.000169541831995934,1.66219792640896e-34],["Ama","Amy-d","Amy-p","Argk","Cyt-b5-r","Eip74EF","grk","Hsp23","Hsp26","Hsp27","NimC3","E(spl)m3-HLH","Mlc2","MtnA","nkd","sesB","sev","sn",null,"Treh","CrebA","Galphaf","Sema2a","alpha-Est1","Reg-2","stet","Sr-CIII","teq","Mct1","Sox100B","Picot","stan",null,"CG3457","Ant2","RhoGAP1A","Idgf4","SPARC","l(1)G0196","bgm",null,null,"Ercc1","NimB4","NimC2","miple2",null,"Hml","CG3168","CG12116","spirit","CG2444","CG9503","CG9164","wus","CCKLR-17D3","bves","CG4297","CG5001","CG5397",null,"Sr-CIV","CG8852","Cyp28d2","Cyp28d1","CG12512",null,"CG8292","firl","CG5418",null,"ChLD3","SPH93",null,"Cyp6w1","Spn42Dc","CG13744","PPO2","CG13743","CG1648","Lime",null,"CG12934","Prip","CG8501",null,"Cyp9h1","sug",null,"CG8468","CG8157","CG4927","CG5550","veil","CG5773","Mctp","CG15080","CG18190",null,"CG9416","twz","wdp","gas",null,"CG3831","Eglp4","Sesn","CG18128","Ance-5","CG3770","Gale","CG9119","Gk2","Cyp4d20","Ctl1","slow","SP1173","CG10063",null,"mthl6","CG13308","CG5144","Slc45-1","CG4476","Muc68Ca","CG10663","Tgi","CG5895","CG5151","MYPT-75D","CG3961","Gbs-76A","zye","CG10581","olf413",null,"CG1124","CG9837","CAHbeta","CG16749","CG8500","GILT1","Npc2b","Pde6","CG14856",null,"EndoU","CG5246","CG14326","pasi1","PKD","CG15695","cDIP","CG6763","CG10184","CG10365","Kal1","CG6356","tnc","CG12290","CG6330","CG14257","CG17189","Sid","CG14516","Cyt-c1L","CG9733","Rpt6R","CG12071","CG1544","Ugt317A1","lectin-28C","CG2854","CG13315","cpx","cbt","CG12880","CG30088","CG30148","Mal-A5","Pgant9","LpR2","CG31125","SKIP",null,"ATPsynepsilonL","CG31642",null,"edin","CG32313","CG32407","CG32625","X11Lbeta","CG32681","CG33307","CG33346","CG33643","CG34057","CG31808","GstE2","CG6006","CG32335","CG34305","CG34330",null,"bbg",null,"CG34434","CG42324","comm3","DIP-epsilon","CG42369","CG42394","CG42395","CG42458","NimC4","CG42524","Pif1A","bou",null,null,"ps"],["40831","36932","47764","39041","35008","39962","34171","39077","39075","39078","48772","43156","43587","41202","44843","32007","32039","31717",null,"45368","39682","39861","36846","40909","38075","38169","44382","39048","31198","45039","36865","36125",null,"31196","32008","5740235","31926","43230","33137","44117",null,null,"36654","34814","34818","44787",null,"39529","31612","31782","31797","32121","32424","32467","32683","32864","33083","33222","33308","33313",null,"33576","33577","33748","33749","33766",null,"34125","34627","34663",null,"35002","35049",null,"35543","35600","35906","35910","35911","36009","36023",null,"36124","36237","36347",null,"36412","36424",null,"36577","36700","36852","36883","36968","37053","37165","37167","37179",null,"37223","37456","37548","37572",null,"37648","37739","37755","37756","37980","37991","38076","38124","38221","38311","38523","38540","38744","38764",null,"38839","39011","39040","39055","39056","2768980","39421","39521","39752","39772","40032","40056","40102","40255","40326","40453",null,"40613","41057","41070","41111","41203","41650","41710","41760","41790",null,"41931","42072","42120","42139","42203","42468","42512","42754","42783","42801","42865","42893","42990","43182","43240","43257","43263","43390","43444","43457","43601","43635","43660","43689","37590","53542","31219","50300","64877","33224","43328","246447","37318","35828","246627","43105","318599","42601",null,"326144","326149",null,"317900","317973","38667","318128","31997","318151","2768914","2768690","3772011","3885645","10178782","37107","41962","261663","5740825","2768869",null,"39583",null,"31472","38430","39696","7354433","33894","7354424","7354436","2768945","34824","8674084","8674049","31644",null,null,"44258"]],"container":"<table class=\"display\">\n  <thead>\n    <tr>\n      <th> <\/th>\n      <th>baseMean<\/th>\n      <th>log2FoldChange<\/th>\n      <th>lfcSE<\/th>\n      <th>stat<\/th>\n      <th>pvalue<\/th>\n      <th>padj<\/th>\n      <th>symbol<\/th>\n      <th>entrez<\/th>\n    <\/tr>\n  <\/thead>\n<\/table>","options":{"columnDefs":[{"className":"dt-right","targets":[1,2,3,4,5,6]},{"orderable":false,"targets":0}],"order":[],"autoWidth":false,"orderClasses":false}},"evals":[],"jsHooks":[]}</script>
</div>
</div>
</section>
<section id="plotting-of-deseq2-results" class="level3" data-number="5.5">
<h3 data-number="5.5" class="anchored" data-anchor-id="plotting-of-deseq2-results"><span class="header-section-number">5.5</span> Plotting of DESEQ2 results</h3>
<section id="plotcounts" class="level4" data-number="5.5.1">
<h4 data-number="5.5.1" class="anchored" data-anchor-id="plotcounts"><span class="header-section-number">5.5.1</span> plotCounts</h4>
<p>To visualize the counts for a particular gene</p>
<div class="cell">
<details>
<summary>Code</summary>
<div class="sourceCode cell-code" id="cb25"><pre class="sourceCode numberSource r number-lines code-with-copy"><code class="sourceCode r"><span id="cb25-1"><a href="#cb25-1"></a><span class="co">#getting the top gene</span></span>
<span id="cb25-2"><a href="#cb25-2"></a>top <span class="ot">&lt;-</span> <span class="fu">rownames</span>(res)[<span class="fu">which.min</span>(res<span class="sc">$</span>padj)]</span>
<span id="cb25-3"><a href="#cb25-3"></a>top_gene<span class="ot">&lt;-</span>res<span class="sc">$</span>symbol[[top]] </span>
<span id="cb25-4"><a href="#cb25-4"></a><span class="co">#plotCounts for the data object </span></span>
<span id="cb25-5"><a href="#cb25-5"></a><span class="fu">plotCounts</span>(dds_seq, <span class="at">gene =</span> top, <span class="at">intgroup =</span> <span class="st">"condition"</span>, <span class="at">main =</span> top_gene) </span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</details>
<div class="cell-output-display">
<div id="fig-plotCounts" class="quarto-figure quarto-figure-center anchored">
<figure class="figure">
<p><img src="rna_seq_files/figure-html/fig-plotCounts-1.png" class="img-fluid figure-img" width="672"></p>
<p></p><figcaption aria-hidden="true" class="figure-caption">Figure 3: plotCounts of top gene</figcaption><p></p>
</figure>
</div>
</div>
</div>
<p>In <a href="#fig-plotCounts">Figure&nbsp;3</a>, We can visualize the normalized counts over both conditions. It helps us to infer the level of difference in gene expression in different conditions. Hence , we can say if there is no or up or down regulation of a gene in a particular condition.</p>
</section>
<section id="plotma" class="level4" data-number="5.5.2">
<h4 data-number="5.5.2" class="anchored" data-anchor-id="plotma"><span class="header-section-number">5.5.2</span> plotMA</h4>
<p>An MA-plot provides a useful overview for an experiment with a two-group comparison.</p>
<div class="cell">
<details>
<summary>Code</summary>
<div class="sourceCode cell-code" id="cb26"><pre class="sourceCode numberSource r number-lines code-with-copy"><code class="sourceCode r"><span id="cb26-1"><a href="#cb26-1"></a><span class="fu">plotMA</span>(res)</span>
<span id="cb26-2"><a href="#cb26-2"></a><span class="fu">abline</span>(<span class="at">h=</span><span class="fu">c</span>(<span class="sc">-</span><span class="dv">1</span>,<span class="dv">1</span>), <span class="at">col=</span><span class="st">"dodgerblue"</span>, <span class="at">lwd=</span><span class="dv">2</span>)</span>
<span id="cb26-3"><a href="#cb26-3"></a><span class="fu">legend</span>(<span class="fl">1e+04</span>, <span class="dv">3</span>, <span class="at">legend=</span><span class="fu">c</span>(<span class="st">"Non Sign"</span>, <span class="st">"Sign"</span>), </span>
<span id="cb26-4"><a href="#cb26-4"></a>       <span class="at">fill =</span> <span class="fu">c</span>(<span class="st">"grey"</span>,<span class="st">"blue"</span>)) </span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</details>
<div class="cell-output-display">
<div id="fig-plotMA" class="quarto-figure quarto-figure-center anchored">
<figure class="figure">
<p><img src="rna_seq_files/figure-html/fig-plotMA-1.png" class="img-fluid figure-img" width="672"></p>
<p></p><figcaption aria-hidden="true" class="figure-caption">Figure 4: An MA-plot of changes induced by treatment.</figcaption><p></p>
</figure>
</div>
</div>
</div>
<p>In <a href="#fig-plotMA">Figure&nbsp;4</a> ,The log2 fold change for a particular comparison is plotted on the y-axis and the average of the counts normalized by size factor is shown on the x-axis Each gene is represented with a dot. Genes with an adjusted p value below a threshold (here 0.1, the default) are shown in blue.</p>
</section>
<section id="volcanoplot" class="level4" data-number="5.5.3">
<h4 data-number="5.5.3" class="anchored" data-anchor-id="volcanoplot"><span class="header-section-number">5.5.3</span> Volcanoplot</h4>
<p>A volcano plot is a type of scatterplot that shows statistical significance (P value) versus magnitude of change (fold change). It enables quick visual identification of genes with large fold changes that are also statistically significant. These may be the most biologically significant genes. In a volcano plot, the most up-regulated genes are towards the right, the most down-regulated genes are towards the left, and the most statistically significant genes are towards the top.</p>
<div class="cell">
<details>
<summary>Code</summary>
<div class="sourceCode cell-code" id="cb27"><pre class="sourceCode numberSource r number-lines code-with-copy"><code class="sourceCode r"><span id="cb27-1"><a href="#cb27-1"></a><span class="fu">EnhancedVolcano</span>(res,</span>
<span id="cb27-2"><a href="#cb27-2"></a>                <span class="at">lab =</span> res<span class="sc">$</span>symbol,</span>
<span id="cb27-3"><a href="#cb27-3"></a>                <span class="at">x =</span> <span class="st">'log2FoldChange'</span>,</span>
<span id="cb27-4"><a href="#cb27-4"></a>                <span class="at">y =</span> <span class="st">'pvalue'</span>,</span>
<span id="cb27-5"><a href="#cb27-5"></a>                <span class="at">pCutoff =</span> <span class="fl">10e-32</span>,</span>
<span id="cb27-6"><a href="#cb27-6"></a>                <span class="at">FCcutoff =</span> <span class="dv">2</span>,</span>
<span id="cb27-7"><a href="#cb27-7"></a>                <span class="at">pointSize =</span> <span class="fl">3.0</span>,</span>
<span id="cb27-8"><a href="#cb27-8"></a>                <span class="at">labSize =</span> <span class="fl">6.0</span> ,</span>
<span id="cb27-9"><a href="#cb27-9"></a>                )</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</details>
<div class="cell-output-display">
<div id="fig-volcano" class="quarto-figure quarto-figure-center anchored">
<figure class="figure">
<p><img src="rna_seq_files/figure-html/fig-volcano-1.png" class="img-fluid figure-img" width="672"></p>
<p></p><figcaption aria-hidden="true" class="figure-caption">Figure 5: Volcano Plot of Differentially Expressed genes</figcaption><p></p>
</figure>
</div>
</div>
</div>
<p>In <a href="#fig-volcano">Figure&nbsp;5</a>, differentially expressed genes which are statistically significant are labelled in red color. The pCutoff and FCcutoff can also be changed based on the needs.</p>
</section>
<section id="heatmap-1" class="level4" data-number="5.5.4">
<h4 data-number="5.5.4" class="anchored" data-anchor-id="heatmap-1"><span class="header-section-number">5.5.4</span> Heatmap</h4>
<p>Since the clustering is only relevant for genes that actually carry a signal, one usually would only cluster a subset of the most highly variable genes. Here, I selected the 20 genes with the highest variance across samples. Working with the rlog transformed counts:</p>
<div class="cell">
<details>
<summary>Code</summary>
<div class="sourceCode cell-code" id="cb28"><pre class="sourceCode numberSource r number-lines code-with-copy"><code class="sourceCode r"><span id="cb28-1"><a href="#cb28-1"></a><span class="co">#ordering the genes based on l2fc,retrieving the first 20 gene rownames</span></span>
<span id="cb28-2"><a href="#cb28-2"></a>genes <span class="ot">&lt;-</span> <span class="fu">order</span>(res_lfc<span class="sc">$</span>log2FoldChange, <span class="at">decreasing=</span><span class="cn">TRUE</span>)[<span class="dv">1</span><span class="sc">:</span><span class="dv">20</span>]</span>
<span id="cb28-3"><a href="#cb28-3"></a>gene_names<span class="ot">&lt;-</span><span class="fu">rownames</span>(res_lfc)[genes]</span>
<span id="cb28-4"><a href="#cb28-4"></a></span>
<span id="cb28-5"><a href="#cb28-5"></a><span class="co">#getting the gene symbols of those 20 rownames</span></span>
<span id="cb28-6"><a href="#cb28-6"></a>names<span class="ot">&lt;-</span>res_lfc<span class="sc">$</span>symbol[gene_names]</span>
<span id="cb28-7"><a href="#cb28-7"></a><span class="co">#making it into a dataframe</span></span>
<span id="cb28-8"><a href="#cb28-8"></a>names<span class="ot">&lt;-</span><span class="fu">data.frame</span>(names)</span>
<span id="cb28-9"><a href="#cb28-9"></a></span>
<span id="cb28-10"><a href="#cb28-10"></a><span class="co"># to see the actual data,the fragment counts,using the assay function.</span></span>
<span id="cb28-11"><a href="#cb28-11"></a>data<span class="ot">&lt;-</span><span class="fu">colnames</span>(<span class="fu">assay</span>(rld)[gene_names,])</span>
<span id="cb28-12"><a href="#cb28-12"></a><span class="co">#cbinding the gene symbols along with assay(rld)</span></span>
<span id="cb28-13"><a href="#cb28-13"></a>new<span class="ot">&lt;-</span><span class="fu">cbind</span>(names,<span class="fu">assay</span>(rld)[gene_names,])</span>
<span id="cb28-14"><a href="#cb28-14"></a>df <span class="ot">&lt;-</span> <span class="fu">subset</span> (new, <span class="at">select =</span> <span class="sc">-</span>names)</span>
<span id="cb28-15"><a href="#cb28-15"></a></span>
<span id="cb28-16"><a href="#cb28-16"></a><span class="co">#for annotations on condition and sequencing type</span></span>
<span id="cb28-17"><a href="#cb28-17"></a>annot_col <span class="ot">&lt;-</span> samples_initial<span class="sc">%&gt;%</span></span>
<span id="cb28-18"><a href="#cb28-18"></a>  <span class="fu">column_to_rownames</span>(<span class="st">'sample'</span>) <span class="sc">%&gt;%</span></span>
<span id="cb28-19"><a href="#cb28-19"></a>  dplyr<span class="sc">::</span><span class="fu">select</span>(condition,type) <span class="sc">%&gt;%</span></span>
<span id="cb28-20"><a href="#cb28-20"></a>  <span class="fu">as.data.frame</span>()</span>
<span id="cb28-21"><a href="#cb28-21"></a></span>
<span id="cb28-22"><a href="#cb28-22"></a><span class="co">#using pheatmap for heatmap</span></span>
<span id="cb28-23"><a href="#cb28-23"></a><span class="fu">pheatmap</span>(df, <span class="at">cluster_rows=</span><span class="cn">TRUE</span>, <span class="at">labels_row =</span> names<span class="sc">$</span>names,<span class="at">legend_breaks =</span> <span class="sc">-</span><span class="dv">2</span><span class="sc">:</span><span class="dv">2</span>,</span>
<span id="cb28-24"><a href="#cb28-24"></a>         <span class="at">cluster_cols=</span><span class="cn">FALSE</span>, <span class="at">annotation_col=</span>annot_col, <span class="at">main=</span> <span class="st">"Heatmap of expression of genes"</span>,<span class="at">scale =</span> <span class="st">"row"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</details>
<div class="cell-output-display">
<div id="fig-heatmap" class="quarto-figure quarto-figure-center anchored">
<figure class="figure">
<p><img src="rna_seq_files/figure-html/fig-heatmap-1.png" class="img-fluid figure-img" width="672"></p>
<p></p><figcaption aria-hidden="true" class="figure-caption">Figure 6: Heatmap of relative rlog-transformed values across samples</figcaption><p></p>
</figure>
</div>
</div>
</div>
<p>In <a href="#fig-heatmap">Figure&nbsp;6</a>, annotation information are shown with colored bars at the top. It helps us to find out the over expression or under expression of top selected genes in different samples. The clustering of gene expression can be studied using the annotations provided.</p>
</section>
</section>
</section>

</main>
<!-- /main column -->
<script id="quarto-html-after-body" type="application/javascript">
window.document.addEventListener("DOMContentLoaded", function (event) {
  const icon = "";
  const anchorJS = new window.AnchorJS();
  anchorJS.options = {
    placement: 'right',
    icon: icon
  };
  anchorJS.add('.anchored');
  const clipboard = new window.ClipboardJS('.code-copy-button', {
    target: function(trigger) {
      return trigger.previousElementSibling;
    }
  });
  clipboard.on('success', function(e) {
    // button target
    const button = e.trigger;
    // don't keep focus
    button.blur();
    // flash "checked"
    button.classList.add('code-copy-button-checked');
    var currentTitle = button.getAttribute("title");
    button.setAttribute("title", "Copied!");
    setTimeout(function() {
      button.setAttribute("title", currentTitle);
      button.classList.remove('code-copy-button-checked');
    }, 1000);
    // clear code selection
    e.clearSelection();
  });
  function tippyHover(el, contentFn) {
    const config = {
      allowHTML: true,
      content: contentFn,
      maxWidth: 500,
      delay: 100,
      arrow: false,
      appendTo: function(el) {
          return el.parentElement;
      },
      interactive: true,
      interactiveBorder: 10,
      theme: 'quarto',
      placement: 'bottom-start'
    };
    window.tippy(el, config); 
  }
  const noterefs = window.document.querySelectorAll('a[role="doc-noteref"]');
  for (var i=0; i<noterefs.length; i++) {
    const ref = noterefs[i];
    tippyHover(ref, function() {
      let href = ref.getAttribute('href');
      try { href = new URL(href).hash; } catch {}
      const id = href.replace(/^#\/?/, "");
      const note = window.document.getElementById(id);
      return note.innerHTML;
    });
  }
  var bibliorefs = window.document.querySelectorAll('a[role="doc-biblioref"]');
  for (var i=0; i<bibliorefs.length; i++) {
    const ref = bibliorefs[i];
    const cites = ref.parentNode.getAttribute('data-cites').split(' ');
    tippyHover(ref, function() {
      var popup = window.document.createElement('div');
      cites.forEach(function(cite) {
        var citeDiv = window.document.createElement('div');
        citeDiv.classList.add('hanging-indent');
        citeDiv.classList.add('csl-entry');
        var biblioDiv = window.document.getElementById('ref-' + cite);
        if (biblioDiv) {
          citeDiv.innerHTML = biblioDiv.innerHTML;
        }
        popup.appendChild(citeDiv);
      });
      return popup.innerHTML;
    });
  }
});
</script>
</div> <!-- /content -->



</body></html>